### DISCUSSION TITLE
NewRelic segfaults with Revolt and they avoid fixing it

### DISCUSSION THREAD

**Discussion by ostrolucky** [revoltphp/event-loop#69] [General]
Following simple code segfaults with NR php extension enabled
```php
<?php
use Revolt\EventLoop;

require_once 'vendor/autoload.php';
EventLoop::defer(function () {});
EventLoop::run();
```

I have created support request on NR side and they responded with

> Hi Gabriel,
> 
> Thanks for that update! We were able to look further into the Revolt framework and it appears it is a concurrent PHP framework, which is not currently supported by the agent.  This can be seen in the PHP agent's compatibility and requirements documentation at the following link:
> 
> - https://docs.newrelic.com/docs/apm/agents/php-agent/getting-started/php-agent-compatibility-requirements/
> 
> To clarify, the PHP agent also does not support any Apache multi-processing model (MPM) that involves threading.  The PHP language development team discourages the use of a threaded MPM with Apache, and we have chosen not to implement support for threaded MPMs in the PHP agent for similar stability reasons. This includes the worker and event MPMs.  We also have documentation outlining this at the following link:
> 
> - https://docs.newrelic.com/docs/apm/agents/php-agent/troubleshooting/threaded-apache-worker-mpms/
> 
> One potential workaround would be to instead use the [PHP OpenTelemetry solution](https://github.com/open-telemetry/opentelemetry-php) to send asynchronous data to the New Relic One. Besides our commitment to being [supportive](https://docs.newrelic.com/docs/integrations/open-source-telemetry-integrations/opentelemetry/introduction-opentelemetry-new-relic/) of OpenTelemetry, it’s also an advantage to you in that you won’t have to instrument with our specific code, but by using OTEL that you can future proof by putting the end point agnostic instrumentation calls in your code.
> 
> When using PHP OTEL, each span keeps its own state information (including parent information) without any global variables. So if someone makes a span in one thread, they can also start/stop/pause it independently of other spans. This is different from the PHP agent, where doing so would keep overwriting the global transaction information.  With this option, you can keep track of spans and set them active as needed when switching threads while also keeping track of their timing so you don’t time things when the thread is paused.
> 
> For more information on OpenTelemetry, I'd recommend viewing our introduction documentation at the following link:
> 
> https://docs.newrelic.com/docs/more-integrations/open-source-telemetry-integrations/opentelemetry/opentelemetry-introduction/ 
> 
> Although the likelihood of this option being implemented is small, I can file a feature request on your behalf to have support for this framework to be potentially added in a future release if you like.  Would you like for me to go ahead and do that for you?
> 
> Thank you,
> 
> Tobby Lugo
> Sr. Technical Support Engineer
> New Relic

Can you advice what should I respond them with? I think they are not correct in claiming Revolt is a concurrent PHP framework and they should support this. I think it's in your interest as well that they fix this properly.

**Comment by bwoebi** [comment]
I suppose, what they are referring to is in particular the support of _concurrent open spans_. I.e. they are likely tracking the current active span via function start and end. However when a fiber switches its stack, the open span must no longer be the last started function, but the last started function on the specific target fiber stack.
This is likely a core assumption of quite some internal structures of their extension and may not be a trivial change.

At Datadog (I'm working on the Datadog PHP Tracer), we had that same issue, that we crashed with swapped fibers, for that very reason, which was fixed in our latest release by tracking the active span with respect to the active fibers.
